sdks/router-sdk/src/utils/encodeMixedRouteToPath.ts

```diff
@@ -1,48 +1,103 @@
 import { pack } from '@ethersproject/solidity'
-import { Currency } from '@uniswap/sdk-core'
+import { Currency, Token } from '@uniswap/sdk-core'
+import { Pair } from '@uniswap/v2-sdk'
 import { Pool as V3Pool } from '@uniswap/v3-sdk'
 import { Pool as V4Pool } from '@uniswap/v4-sdk'
+import {
+  ADDRESS_ZERO,
+  MIXED_QUOTER_V2_V2_FEE_PATH_PLACEHOLDER,
+  MIXED_QUOTER_V2_V3_FEE_PATH_PLACEHOLDER,
+  MIXED_QUOTER_V2_V4_FEE_PATH_PLACEHOLDER,
+  MIXED_QUOTER_V1_V2_FEE_PATH_PLACEHOLDER,
+} from '../constants'
 import { MixedRouteSDK } from '../entities/mixedRoute/route'
-import { V2_FEE_PATH_PLACEHOLDER } from '../constants'
-import { TPool } from './TPool'
 
+type TPool = Pair | V3Pool | V4Pool
 /**
  * Converts a route to a hex encoded path
  * @notice only supports exactIn route encodings
  * @param route the mixed path to convert to an encoded path
  * @returns the exactIn encoded path
  */
 export function encodeMixedRouteToPath(route: MixedRouteSDK<Currency, Currency>): string {
-  const firstInputToken: Currency = route.input.wrapped
-
-  const { path, types } = route.pools.reduce(
-    (
-      { inputToken, path, types }: { inputToken: Currency; path: (string | number)[]; types: string[] },
-      pool: TPool,
-      index
-    ): { inputToken: Currency; path: (string | number)[]; types: string[] } => {
-      if (pool instanceof V4Pool) throw 'Encoding mixed routes with V4 not supported'
-      const outputToken: Currency = pool.token0.equals(inputToken) ? pool.token1 : pool.token0
-      if (index === 0) {
-        return {
-          inputToken: outputToken,
-          types: ['address', 'uint24', 'address'],
-          path: [
-            inputToken.wrapped.address,
-            pool instanceof V3Pool ? pool.fee : V2_FEE_PATH_PLACEHOLDER,
-            outputToken.wrapped.address,
-          ],
-        }
+  const containsV4Pool = route.pools.some((pool) => pool instanceof V4Pool)
+
+  const firstCurrencyIn: Currency = route.input
+  const firstTokenIn: Token = firstCurrencyIn.wrapped
+
+  let path: (string | number)[]
+  let types: string[]
+
+  if (containsV4Pool) {
+    path = [firstTokenIn.isNative ? ADDRESS_ZERO : firstTokenIn.address]
```

Comment:
firstToken in will never be `isNative` because it is wrapped.  I'm afraid with V4 this will take ETH pools and try and trade WETH on them?

Reply:
iirc a route should never have an improper inputToken. it will always be a pool token. Trade is the one that may have a native input, that's wrapped to a token input for the route

Reply:
you can utilize route.adjustedInput

Reply:
aw good point!